File: bindings/dart/lib/src/mcp.dart
import "dart:convert";
import "dart:io";

import "types.dart";

final class NexusClient {
  NexusClient._(this._endpoint);

  final Uri _endpoint;
  int _nextId = 1;

  static Future<NexusClient> connectHttp(String endpoint) async {
    return NexusClient._(Uri.parse(endpoint));
  }

  Future<Object?> send(String method, JsonObject params) async {
    final client = HttpClient()
      ..connectionTimeout = const Duration(seconds: 30);
    try {
      final request = await client.postUrl(_endpoint);
      request.headers.contentType = ContentType.json;
      request.write(
        jsonEncode(
          <String, Object?>{
            "jsonrpc": "2.0",
            "id": _nextId++,
            "method": method,
            "params": params,
          },
        ),
      );
      final response =
          await request.close().timeout(const Duration(seconds: 30));
      final body = jsonDecode(await response.transform(utf8.decoder).join());
      if (body is! JsonObject) {
        throw StateError("MCP response must be a JSON object");
      }
      if (body["error"] != null) {
        throw StateError("MCP request failed: ${jsonEncode(body["error"])}");
      }
      return body["result"];
    } finally {
      client.close(force: true);
    }
  }

  Future<List<JsonObject>> listTools() async {
    final result = await send("tools/list", const <String, Object?>{});
    if (result is! JsonObject || result["tools"] is! List<Object?>) {
      return const <JsonObject>[];
    }
    return (result["tools"]! as List<Object?>)
        .whereType<JsonObject>()
        .map(copyJsonObject)
        .toList(growable: false);
  }

  Future<Object?> invokeTool(String name, JsonObject args) {
    return send(
      "tools/call",
      <String, Object?>{"name": name, "arguments": copyJsonObject(args)},
    );
  }
}

File: bindings/python/orchustr/mcp.py
from __future__ import annotations

import asyncio
import json
import urllib.request


class NexusClient:
    def __init__(self, endpoint: str) -> None:
        self._endpoint = endpoint
        self._next_id = 1

    @classmethod
    async def connect_http(cls, endpoint: str) -> "NexusClient":
        return cls(endpoint)

    async def send(self, method: str, params: dict) -> dict:
        payload = json.dumps(
            {"jsonrpc": "2.0", "id": self._next_id, "method": method, "params": params}
        ).encode("utf-8")
        self._next_id += 1

        def _request() -> dict:
            request = urllib.request.Request(
                self._endpoint,
                data=payload,
                headers={"Content-Type": "application/json"},
                method="POST",
            )
            with urllib.request.urlopen(request, timeout=30) as response:
                body = json.loads(response.read().decode("utf-8"))
                return body["result"]

        return await asyncio.to_thread(_request)

    async def list_tools(self) -> list[dict]:
        return (await self.send("tools/list", {})).get("tools", [])

    async def invoke_tool(self, name: str, args: dict) -> dict:
        return await self.send("tools/call", {"name": name, "arguments": args})
